docs: Polish the langref concepts (#2241)
slint-ui/slint: docs/langref/src/concepts/purity.md

result of the evaluation. When a property changes, all dependent properties get
notified, so that the next time their value is read, their binding is re-evaluated.

For any reactive system to work well, bindings evaluation should be "pure":

> Evaluating a property shouldn't have side-effects on other properties or change

Reviewer: I suppose the "properties" was added to clarification, but just today I was helping an ambassador with their code and in their C+ callback that's declared `pure` they were changing the rows of models. So I'd leave it "open" and not "limited" to properties here.

```suggestion
Evaluating a property shouldn't have side-effects and change
```

> Evaluating a property shouldn't have side-effects nor change
> any observable state but the property itself. Side effects are problematic

Reviewer: Let's use the idiomatic "neither" for the conjunction then.

```suggestion
Evaluating a property should have neither side-effects nor change
any observable state but the property itself. Side effects are problematic
```

That said, I prefer the original way it was written. A change to observable state other than the property itself **is** a side-effect after all, isn't it?

Reviewer: I'm inclined to agree, but "neither ... nor" with two sides that are not actually mutually exclusive is more confusing IMO than helping. How about two sentences?

"Evaluating a property shouldn't have side-effects. An example of a side-effect is when a binding expression not only computes the value of the property it's associated with, but also assigns new values to other properties."

> is "pure", otherwise it's said to have side-effects. Side-effects are problematic

Reviewer: Very nice!!

because it's not always clear when they will happen: Lazy evaluation may change
their order or affect whether they happen at all. In addition, changes to
properties during their binding evaluation due to a side-effect may result in
unexpected behavior.

For this reason, bindings in Slint _must_ be pure. The Slint compiler enforces
code in pure contexts to be free of side effects. Pure contexts include binding
expressions, bodies of pure functions, and bodies of pure callback handlers.
In such a context, it's not allowed to change a property, or call a non-pure
callback or function.

Annotate callbacks and public functions with the `pure` keyword to make them
accessible from property bindings and other pure callbacks and functions.

> The purity of private functions is automatically inferred. Annotated them
> explicitly with `pure` to enforce their purity.

Reviewer: ```suggestion
The purity of private functions is automatically inferred. Annotate them
explicitly with `pure` to enforce their purity.
```
